File: Tools/VfxLab/web/sketches/infinity-castle-shift.js
// Infinity Castle: Shift — castle command proposal for the Infinity Castle kit, not the game.
// Nothing in Source/RimArt draws this yet.
//
// What it is for (agreed in outline 2026-09-23; every number is a placeholder and will be an XML
// field). A command the carrier plays from the biwa room while the castle is open. Every command is
// one strum, and strums are 1.5 s apart. Shift: pick a room (not the biwa room) and one of four
// directions. The room slides straight that way until its wall touches another room, at most 20
// cells. Everything in it rides along: pawns, items, corpses. Its doorways close as the rooms part;
// where it now touches another room along 3 or more cells, a new doorway opens. A room pushed into
// open void has no doorways, so whoever is inside is cut off until another Shift. Void rule: a pawn
// standing in a doorway that breaks is left over the void, drops, and comes up in a random room
// through a floor door, unharmed.
//
// Order (times with the default sliders):
//   0.00  the castle at rest: the biwa room (west) with the carrier on the dais, a corridor, the
//         tatami room A with a raider in it, room N north of A, the hall E east of A across 6 cells
//         of void with a colonist in it, a corridor off E and room D beside that corridor
//   0.20  the bachi lifts; 0.40 strum: rings leave the biwa, and A's outline flashes
//   0.45  A's two doorways (to the corridor and to N) slide shut in 0.15 s
//   0.65  A slides east, speeding up, two afterimages of its outline trailing; the raider rides
//   1.50  A's east wall meets E's west wall: dust along the seam, a small shake, the raider
//         stumbles 0.25 cells; 0.15 s later the new doorway between A and E slides open
//   result: A is joined to E only; the corridor and N keep closed doors facing the void
// Scenarios: "nothing in the way" has no E, so A slides the full 20 cells, eases to a stop and has
// no doorway at all. "raider in the doorway" puts a second raider in A's west doorway: once A has
// moved off, it is over the void, drops, and rises through a floor door in D.
//
// Drawing: the sliding room is drawn with the same meshes as a room at rest, offset (in game the
// real room is hidden under the drawing while it moves and the cells are moved at the stop; riding
// pawns need MimicRender or a fade). Level shapes and stand-in pawns only, no per-facing method.
import { P } from './lib/six-paths-impact.js';
import {
  local, slideDistance, moved, doorChanges, drawVoid, drawRoom, doorway, floorDoor, doorAt, doorEnd, figure, rising, sinking,
  nakime, strum, roomFlash, outline, dustLine, centreOf, cellCentre, seatOf, biwaOf, Enemy, Ally, smooth, clamp, bump, easeOut,
} from './lib/infinity-castle.js';

// Decided values.
const Max = 20, Rise = .55, Door0 = .09;
const Layout = [
  ['biwa', -22, -5, 9, 9], ['corridor', -13, -2, 7, 5], ['tatami', -6, -5, 9, 9], ['tatami', -5, 4, 7, 7],
  ['hall', 9, -8, 11, 11], ['corridor', 12, 3, 5, 9], ['tatami', 5, 4, 7, 7],
];
const A = 2, D = 6, Dir = { x: 1, z: 0 };
const Rider = [-2, 1], Colonist = [13, -3], InDoorway = [-6, 0];

// The castle for a scenario, the slide, and the castle after it; built once each.
const cache = new Map();
function setup(scenario) {
  if (cache.has(scenario)) return cache.get(scenario);
  const specs = scenario === 'nothing in the way' ? Layout.slice(0, 4) : Layout;
  const before = local(specs);
  const { d, blocked } = slideDistance(before.rooms, A, Dir.x, Dir.z, Max);
  const after = moved(before, A, Dir.x * d, Dir.z * d);
  const out = { before, after, d, blocked, changes: doorChanges(before, after) };
  cache.set(scenario, out);
  return out;
}

function times(p) {
  const { d, blocked } = setup(p.scenario);
  const strumAt = p.lead, close0 = strumAt + .05, slide0 = strumAt + .25;
  const slideTime = .25 + d / p.speed, stop = slide0 + slideTime, open = stop + .15;
  // Void rule: the doorway raider drops once A's west wall has moved 0.9 cells off its cell.
  const u = blocked ? Math.pow(.9 / d, 1 / 1.5) : .3;
  const fall = slide0 + u * slideTime, land = fall + .75;
  const last = p.scenario === 'raider in the doorway' ? Math.max(open + .2, land + doorEnd(Rise * .75)) : open + .2;
  return { strumAt, close0, slide0, slideTime, stop, open, fall, land, end: last + p.hold };
}

export default {
  kit: 'Infinity Castle', label: 'Shift (sketch)', scene: false,
  params: {
    scenario: { label: 'Scenario', value: 'touches a room', options: ['touches a room', 'nothing in the way', 'raider in the doorway'], group: 'Showcase' },
    depth: { label: 'Rooms at other depths', value: true, group: 'Showcase' },
    lead: P('Before the strum', .4, .2, 1, .05, 'Timing (s)'),
    speed: P('Slide speed (cells a second)', 8, 4, 20, 1, 'Timing (s)'),
    hold: P('Show the result', 1.2, .3, 3, .1, 'Timing (s)'),
  },
  duration(p) { return times(p).end; },
  phases(p) { const t = times(p); return [
    { name: 'At rest', t: 0 }, { name: 'Strum', t: t.strumAt }, { name: 'Slide', t: t.slide0 }, { name: 'Stop', t: t.stop }, { name: 'Result', t: t.open + .2 },
  ]; },
  events(p) {
    const t = times(p), { blocked } = setup(p.scenario);
    const out = [{ t: t.strumAt, type: 'sound', def: 'RimArt_BiwaStrum' }, { t: t.slide0, type: 'sound', def: 'RimArt_CastleSlide' }];
    if (blocked) out.push({ t: t.stop, type: 'shake', value: .03 }, { t: t.stop, type: 'sound', def: 'RimArt_CastleThud' });
    return out;
  },

  draw(s, p, { origin, scene }) {
    const t = times(p);
    if (s < 0 || s >= t.end) return;
    const sun = scene?.shadowVector ?? { x: -.45, z: -.32 }, strength = (scene?.sun?.strength ?? .32) * .6;
    const { before, after, d, blocked, changes } = setup(p.scenario);
    const base = { x: origin.x - .5 + 2, z: origin.z - .5 - 1 };
    drawVoid({ x: origin.x, z: origin.z }, s, { seed: 3, depth: p.depth, reach: 50 });

    // How far A has slid: speeds up into a wall it hits, eases to a stop in open void.
    const u = clamp((s - t.slide0) / t.slideTime);
    const slid = s < t.slide0 ? 0 : d * (blocked ? Math.pow(u, 1.5) : smooth(u));
    const offsets = new Map([[A, { x: Dir.x * slid, z: Dir.z * slid }]]);
    const room = before.rooms[A];
    before.rooms.forEach(r => drawRoom(r, centreOf(base, r, offsets.get(r.id)), { s }));

    // Afterimages trail the room while it moves fast.
    if (s > t.slide0 && s < t.stop + .15) for (const [lag, a] of [[.05, .22], [.1, .11]]) {
      const lu = clamp((s - lag - t.slide0) / t.slideTime), back = d * (blocked ? Math.pow(lu, 1.5) : smooth(lu));
      const fade = 1 - clamp((s - t.stop) / .15);
      outline(room, centreOf(base, room, { x: Dir.x * back, z: Dir.z * back }), a * fade * clamp(u * 4));
    }

    // Doorways: kept ones stay open, broken ones shut before the slide, new ones open after the stop.
    const shut = 1 - smooth((s - t.close0) / .15);
    changes.kept.forEach(dr => doorway(`shift kept ${dr.key}`, dr, base, 1, { offsets }));
    changes.broken.forEach(dr => doorway(`shift broken ${dr.key}`, dr, base, shut, { offsets }));
    if (s >= t.open) changes.made.forEach(dr => doorway(`shift made ${dr.key}`, dr, base, easeOut((s - t.open) / .2)));

    // The stop against E: dust along the seam where the walls met.
    if (blocked) {
      const seamX = base.x + room.x + room.w + d, z0 = base.z + Math.max(room.z, after.rooms[4].z), z1 = base.z + Math.min(room.z + room.h, after.rooms[4].z + after.rooms[4].h);
      dustLine('shift seam', { x: seamX, z: z0 + .5 }, { x: seamX, z: z1 - .5 }, s - t.stop, { count: 12, alpha: .55, spread: .6 });
    }

    // The raider in A rides along and stumbles at the stop; the colonist waits in E.
    const stumble = blocked ? .25 * bump(clamp((s - t.stop) / .3)) : 0;
    const rider = cellCentre(base, Rider[0], Rider[1]);
    figure({ x: rider.x + Dir.x * (slid + stumble), z: rider.z + Dir.z * (slid + stumble) }, Enemy, sun, strength);
    if (p.scenario !== 'nothing in the way') figure(cellCentre(base, Colonist[0], Colonist[1]), Ally, sun, strength);

    // Void rule: the raider in A's west doorway is left over the void, drops, comes up in D.
    if (p.scenario === 'raider in the doorway') {
      const at = cellCentre(base, InDoorway[0], InDoorway[1]);
      if (s < t.fall) figure(at, Enemy, sun, strength);
      else {
        const fu = clamp((s - t.fall) / .5);
        if (fu < 1) figure({ x: at.x, z: at.z - .3 * smooth(fu) }, Enemy, sun, strength * (1 - fu), { scale: 1 - .6 * smooth(fu), dark: smooth(fu) * .9, alpha: 1 - smooth((fu - .7) / .3) });
      }
      const dr = before.rooms[D], landAt = cellCentre(base, dr.x + Math.floor(dr.w / 2), dr.z + Math.floor(dr.h / 2));
      const age = s - t.land, st = doorAt(age, Rise * .75);
      if (age < doorEnd(Rise * .75)) floorDoor('shift void land', landAt, st.open, st.alpha, { s });
      if (age >= Door0) rising('shift void rise', landAt, Enemy, clamp((age - Door0) / Rise), sun, strength);
    }

    // The carrier plays; the room she moves flashes.
    const seat = seatOf(base, before.rooms[0]);
    nakime('shift carrier', seat, sun, strength, { strum: s - t.strumAt > -.25 ? s - t.strumAt : null });
    strum('shift strum', biwaOf(seat), s - t.strumAt, { reach: 5, life: .6 });
    roomFlash(room, centreOf(base, room, offsets.get(A)), s - t.strumAt - .05);
  },
};

File: Source/RimArt/InfinityCastle/Kit/InfinityCastleShift.cs
using System.Collections.Generic;
using System.Linq;
using HarmonyLib;
using RimWorld;
using UnityEngine;
using Verse;
using static RimArt.CastleEffectGraphics;

namespace RimArt
{
    /// <summary>
    /// Shift, the castle command: a room slides straight until its wall touches another room, at most
    /// shiftMaxCells, with everything in it riding along. Its doorways close as the rooms part and a new
    /// doorway opens where it now touches a room along 3 or more cells.
    ///
    /// The cells move once, at the strum, and only the picture slides: the room's floor, walls and
    /// lanterns are rebuilt at the destination at once (with the room's things and pawns), and for the
    /// slide's length the room is drawn on its own, from where it was to where it is, with its riders
    /// drawn behind their real cells by the same offset (<see cref="Patch_PawnRenderer_CastleRide"/>) and
    /// stunned so nobody walks off a room whose picture has not arrived. The picture runs on the map's
    /// own clock, real seconds whether or not the game is paused, as the lanterns do: a command is played
    /// from a paused game as often as not. A save made mid-slide loads with the move done and the
    /// picture skipped. The one thing left until the stop is the wall in the far
    /// cell of each new doorway, so the neighbours cannot step into a doorway before it is drawn open.
    ///
    /// The Void rule: a pawn standing in either cell of a doorway that breaks is left over the void. It
    /// drops, and comes up in a random room through a floor door, unharmed but stunned a moment.
    ///
    /// Numbers are on the AG_InfinityCastle def (<see cref="InfinityCastleRules"/>); the timings that
    /// shape the picture are the Shift sketch's (Tools/VfxLab/web/sketches/infinity-castle-shift.js).
    /// </summary>
    public sealed class CastleShift : IExposable
    {
        public int room, dx, dz, distance;
        /// <summary>When the strum was, on the castle map's clock (real seconds since the castle opened).</summary>
        public float startAt;
        public bool blocked;
        /// <summary>Where the room was, so a loaded castle can still finish the stop's bookkeeping.</summary>
        public int oldX, oldZ;
        /// <summary>The far cell of each new doorway: its wall comes out at the stop.</summary>
        public List<IntVec3> madeCells = new List<IntVec3>();
        public List<Pawn> riders = new List<Pawn>();
        public bool stopped;

        // The picture, not saved: a castle loaded mid-slide finishes without it.
        internal List<CastleDoorway> broken = new List<CastleDoorway>(), made = new List<CastleDoorway>();
        internal CastleRoomGraphics.CastleBatch movingRoom;
        internal (Vector2 a, Vector2 b)? seam;
        internal bool picture, shook;

        // The Shift sketch's order, in seconds from the strum.
        public const float Close0 = 0.05f, Slide0 = 0.25f, CloseFor = 0.15f, OpenAfter = 0.15f, OpenFor = 0.2f, DustFor = 0.7f;

        public float SlideTime => 0.25f + distance / InfinityCastleRules.Of.shiftCellsPerSecond;
        public float StopAt => Slide0 + SlideTime;
        public float OpenAt => StopAt + OpenAfter;
        public float EndAt => OpenAt + OpenFor + DustFor;

        public float AgeAt(float now) => now - startAt;

        /// <summary>How far the room has slid at <paramref name="t"/>: speeding up into a wall it hits, easing to a stop in open void.</summary>
        public float Slid(float t)
        {
            float u = Clamp((t - Slide0) / SlideTime);
            return distance * (blocked ? Mathf.Pow(u, 1.5f) : VfxMath.Smooth(u));
        }

        /// <summary>The picture's offset from the real cells at <paramref name="t"/>: zero once the room has arrived.</summary>
        public Vector2 Behind(float t)
        {
            float back = Slid(t) - distance;
            return new Vector2(dx * back, dz * back);
        }

        public void ExposeData()
        {
            Scribe_Values.Look(ref room, "room");
            Scribe_Values.Look(ref dx, "dx");
            Scribe_Values.Look(ref dz, "dz");
            Scribe_Values.Look(ref distance, "distance");
            Scribe_Values.Look(ref startAt, "startAt");
            Scribe_Values.Look(ref blocked, "blocked");
            Scribe_Values.Look(ref oldX, "oldX");
            Scribe_Values.Look(ref oldZ, "oldZ");
            Scribe_Values.Look(ref stopped, "stopped");
            Scribe_Collections.Look(ref madeCells, "madeCells", LookMode.Value);
            Scribe_Collections.Look(ref riders, "riders", LookMode.Reference);
            if (Scribe.mode == LoadSaveMode.PostLoadInit)
            {
                madeCells = madeCells ?? new List<IntVec3>();
                riders = riders ?? new List<Pawn>();
                riders.RemoveAll(p => p == null);
                picture = false;
                startAt = -1000f;                                   // long over: the stop happens on the first update
            }
        }
    }

    /// <summary>
    /// Crush, the castle command: a room's four walls slam in and draw back, and everyone in the outer
    /// band of its floor is hit. The hit lands at <see cref="Hit"/> after the strum; the picture (the
    /// band lit, the slabs, dust, splinters, the stun stars) is the Crush sketch's. A castle loaded with a
    /// crush in flight lands the hit at once and skips the picture.
    /// </summary>
    public sealed class CastleCrush : IExposable
    {
        public int room;
        public float startAt;
        public bool hitDone;

        // The Crush sketch's order, in seconds from the strum.
        public const float Mark = 0.05f, Telegraph = 0.3f, SlamFor = 0.12f, HoldFor = 0.3f, BackFor = 0.8f;
        public const float Slam = Mark + Telegraph, Hit = Slam + SlamFor, Back = Hit + HoldFor, Open = Back + BackFor;
        /// <summary>Over when the walls are back and the last stun star has gone (1 s from the hit).</summary>
        public const float Length = Open + 0.2f;

        public float AgeAt(float now) => now - startAt;

        /// <summary>How far in the walls are at <paramref name="t"/>, 0..1: speeding up into the hit, a short hold, easing back.</summary>
        public static float In(float t)
        {
            if (t < Slam) return 0f;
            if (t < Hit) return Mathf.Pow((t - Slam) / SlamFor, 2f);
            if (t < Back) return 1f;
            return 1f - VfxMath.Smooth((t - Back) / BackFor);
        }

        public void ExposeData()
        {
            Scribe_Values.Look(ref room, "room");
            Scribe_Values.Look(ref startAt, "startAt");
            Scribe_Values.Look(ref hitDone, "hitDone");
            if (Scribe.mode == LoadSaveMode.PostLoadInit) startAt = -1000f;
        }
    }

    /// <summary>A Seal or Open being drawn: which doorway, which way, and when the strum was. Not saved; a loaded seal is simply shut.</summary>
    internal sealed class CastleSealAnim
    {
        public string key;
        public CastleDoorway door;
        public bool sealing;
        public float startAt;
        /// <summary>Both pictures are over within this: the bar's 0.25 s from 0.2 s, plus the outline's fade.</summary>
        public const float Length = 0.6f;
    }

    /// <summary>
    /// A pawn going through the floor: Drop (the carrier picks the pawn and the room) and the Void rule
    /// (a random room). The move is at once; the picture is the Drop sketch's: a door opens under the
    /// pawn <see cref="doorUnder"/> after the strum, the pawn sinks for 0.4 s and is gone, a door opens
    /// in the far room 0.2 s later and the pawn rises out of it over 0.55 s. While it sinks the pawn is
    /// drawn back at its old cell; in between it is not drawn at all.
    /// </summary>
    public sealed class CastleDrop : IExposable
    {
        public Pawn pawn;
        public IntVec3 from, to;
        public int fromRoom = -1, toRoom = -1;
        public float startAt;
        /// <summary>Seconds after the strum before the door opens: 0.1 for Drop, 0 for the Void rule.</summary>
        public float doorUnder;
        /// <summary>Whether this was a strum (Drop): the rings from the dais and the room's flash.</summary>
        public bool strum;

        public const float Sink = 0.4f, Rise = 0.55f;
        public float SinkStart => doorUnder + DoorThrough;
        public float SinkEnd => SinkStart + Sink;
        public float Arrive => SinkEnd + 0.2f;
        public float AgeAt(float now) => now - startAt;
        public bool DoneAt(float now) => AgeAt(now) > Arrive + DoorEnd(Rise * 0.75f);

        public void ExposeData()
        {
            Scribe_References.Look(ref pawn, "pawn");
            Scribe_Values.Look(ref from, "from");
            Scribe_Values.Look(ref to, "to");
            Scribe_Values.Look(ref fromRoom, "fromRoom", -1);
            Scribe_Values.Look(ref toRoom, "toRoom", -1);
            Scribe_Values.Look(ref startAt, "startAt");
            Scribe_Values.Look(ref doorUnder, "doorUnder");
            Scribe_Values.Look(ref strum, "strum");
            if (Scribe.mode == LoadSaveMode.PostLoadInit) startAt = -1000f;
        }
    }

    /// <summary>
    /// Where the castle's riders are drawn: each pawn's picture offset from its real cell while its room
    /// slides, and the pawns hidden while they fall through the void. Read by the render patch every
    /// frame, written by the castle's map component; static because there is one renderer.
    /// </summary>
    internal static class InfinityCastleRide
    {
        private static readonly Dictionary<Pawn, Vector3> offsets = new Dictionary<Pawn, Vector3>();
        private static readonly HashSet<Pawn> hidden = new HashSet<Pawn>();

        public static int Count => offsets.Count + hidden.Count;
        public static void Ride(Pawn pawn, Vector2 behind) { hidden.Remove(pawn); offsets[pawn] = new Vector3(behind.x, 0f, behind.y); }
        public static void Hide(Pawn pawn) { offsets.Remove(pawn); hidden.Add(pawn); }
        public static void Release(Pawn pawn) { offsets.Remove(pawn); hidden.Remove(pawn); }
        public static void Clear() { offsets.Clear(); hidden.Clear(); }
        public static bool Hidden(Pawn pawn) => hidden.Contains(pawn);
        public static bool Offset(Pawn pawn, out Vector3 offset) => offsets.TryGetValue(pawn, out offset);
    }

    /// <summary>
    /// The game draws every pawn from results computed for where the pawn really is, so a rider is moved
    /// by changing the position every phase is asked for. A falling pawn is not drawn at all.
    /// </summary>
    [HarmonyPatch(typeof(PawnRenderer), nameof(PawnRenderer.DynamicDrawPhaseAt))]
    internal static class Patch_PawnRenderer_CastleRide
    {
        private static bool Prefix(Pawn ___pawn, ref Vector3 drawLoc)
        {
            if (InfinityCastleRide.Count == 0 || ___pawn == null) return true;
            if (InfinityCastleRide.Hidden(___pawn)) return false;
            if (InfinityCastleRide.Offset(___pawn, out Vector3 offset)) drawLoc += offset;
            return true;
        }
    }
}
